shard/src/ext/pug_cog.py

```diff
@@ -0,0 +1,63 @@
+from discord.ext import commands
+from contextlib import suppress
+import time
+import asyncio
+from lib.config import logger
+
+
+class Pug(commands.Cog):
+
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command()
+    async def startpug(self, ctx, requiredPlayers: int = 10):
+        '''
+        Starts a tally for pugs
+        '''
+        settings = self.bot.settings[ctx.guild]
+        pug_emoji = settings.pug_emoji
+
+        t_end = time.time() + 60 * 30
+        user_list = []
+        msg = await ctx.send(f"{requiredPlayers} more {pug_emoji}'s for pugs")
+        await msg.add_reaction(pug_emoji)
+        while time.time() < t_end:
+
+            def check(r, u):
+                return r.message.id == msg.id and str(r.emoji) == pug_emoji
+
+            async def update():
+                num_left = max(0, (requiredPlayers - len(user_list)))
+                await msg.edit(content=f"{num_left} more {pug_emoji}'s for pugs")
+
+            with suppress(TimeoutError):
+                task_add = asyncio.ensure_future(self.bot.wait_for('reaction_add', check=check))
+                task_remove = asyncio.ensure_future(self.bot.wait_for('reaction_remove', check=check))
+                done, _ = await asyncio.wait([task_add, task_remove], timeout=5, return_when=asyncio.FIRST_COMPLETED)
+
+                if task_add in done:
+                    task_remove.cancel()
+                    react, user = task_add.result()
+                    if user and user not in user_list and user != self.bot.user:
+                        user_list.append(user)
+                        t_end += int((settings.pug_timeout_speed / 2) * 60)
+                        await update()
+                elif task_remove in done:
+                    task_add.cancel()
+                    react, user = task_remove.result()
+                    if user and user in user_list and user != self.bot.user:
+                        user_list = [u for u in user_list if u.id != user.id]
+                        await update()
+
```

Comment:
instead of handling both events separately, use `user_list = await react.users().flatten()` from https://discordpy.readthedocs.io/en/latest/api.html#discord.Reaction.users
Not only is it fewer lines, but also it works better when people spam or a bunch of people reacts at once, cause the gateway events can get lost or arrive out of order.

Reply:
ok I wanted to save an API call by not doing this but if you give me the green light on this i'm all in

Reply:
imo it's worth it cause otherwise it's gets out of sync easily